src/operations/scripts.rs

```diff
@@ -0,0 +1,213 @@
+    match mode {
+        // Note: we assume it is "safe" to change the target of an existing
+        // symlink
+        Safe => delete_if_link(dest_path),
+        // Note: we need to delete dest_path for unix::fs::symlink to work
+        // later on
+        Override => delete_if_exists(dest_path),
+    }?;
+    println!("{} -> {}", dest_path.display(), src_path.display());
+    std::os::unix::fs::symlink(src_path, dest_path).map_err(|e| {
+        new_error(&format!(
+            "Could not create link from {} to {}: {}",
+            dest_path.display(),
+            src_path.display(),
+            e
+        ))
+    })
+}
+
+#[cfg(unix)]
+fn delete_if_link(path: &Path) -> Result<(), Error> {
+    // This will make an error if the path does not exist,
+    // but also if we don't have permission to read the path.
+    // In both cases, we want to return early - in the later,
+    // the code will probably fail when creating the symlink
+    // later on.
+    let meta = std::fs::symlink_metadata(path);
+    if meta.is_err() {
+        return Ok(());
+    };
+    let meta = meta.unwrap();
+    if !meta.file_type().is_symlink() {
+        return Err(new_error(&format!(
+            "{} exists and is *not* a symlink",
+            path.display()
+        )));
+    };
+    std::fs::remove_file(path).map_err(|e| {
+        new_error(&format!(
+            "Could not remove existing symlink {}: {}",
+            path.display(),
+            e
+        ))
+    })?;
+    Ok(())
+}
+
+#[cfg(unix)]
+fn delete_if_exists(path: &Path) -> Result<(), Error> {
+    if path.exists() {
+        std::fs::remove_file(path).map_err(|e| {
+            new_error(&format!(
+                "Could not remove existing file {}: {}",
+                path.display(),
+                e
+            ))
+        })?;
+    }
+    Ok(())
+}
+
+fn list_egg_info_dirs(project_path: &Path) -> Result<Vec<PathBuf>, std::io::Error> {
+    let mut res = vec![];
+    for entry in std::fs::read_dir(&project_path)? {
+        let entry = entry?;
+        let path = entry.path();
+        if path.is_dir() {
+            // See https://github.com/rust-lang/rfcs/issues/900
+            // for why we don't use directly path.file_name().ends_with:
+            // OsStr does not have ends_with
+            let file_name = path.file_name().unwrap().to_string_lossy();
+            if file_name.ends_with(".egg-info") {
+                res.push(path);
+            }
+        }
+    }
+    Ok(res)
+}
+
+fn find_egg_info(project_path: &Path) -> Result<PathBuf, Error> {
+    let matches = list_egg_info_dirs(project_path)
+        .map_err(|e| new_error(&format!("While listing project path: {}", e)))?;
+    let num_matches = matches.len();
+    if num_matches != 1 {
+        return Err(new_error(&format!(
+            "Expecting exactly one .egg-info entry, got {}",
+            num_matches
+        )));
+    }
+    Ok(matches[0].clone())
```

Comment:
Unfortunate to do a copy just to return a value!
You could return `matches.swap_remove(0)`, that just moves the match out of the vector in `O(1)`.